AK/Random.cpp

```diff
@@ -1,98 +1,126 @@
-//       See MathObject::random() from LibJS. Where cryptographic security is needed use LibCrypto/SecureRandom.h.
-void fill_with_random([[maybe_unused]] Bytes bytes)
+static inline ErrorOr<void> csprng(void* const buf, size_t size)
 {
 #if defined(AK_OS_SERENITY) || defined(AK_OS_ANDROID) || defined(AK_OS_BSD_GENERIC) || defined(AK_OS_HAIKU) || AK_LIBC_GLIBC_PREREQ(2, 36)
-    arc4random_buf(bytes.data(), bytes.size());
-#elif defined(OSS_FUZZ)
-#else
-    auto fill_with_random_fallback = [&]() {
-        for (auto& byte : bytes)
-            byte = rand();
-    };
+    arc4random_buf(buf, size);
+#elif defined(AK_OS_LINUX)
+    int ret;
+    while (size > 0u) {
+        // The possible errors that can be returned here are:
+        //
+        // EINTR, interrupted by a signal handler.
+        // EAGAIN, the requested entropy wasn't available. Only possible if the GRND_NONBLOCK flag is set, which it isn't.
+        // EINVAL, invalid flags provided, which there aren't.
+        // ENOSYS, the kernel doesn't implement the syscall. The syscall has been available since Linux 3.17, which was released in 2019.
+        // EFAULT, the address is outside the accessible space; the caller is responsible for providing a valid address.
+        //
+        // Therefore, the only somewhat plausible errors are EFAULT and EINTR.
+        // Nonetheless, we will check for the possibility of all the error codes out of an abundance of caution.
+        ret = getrandom(buf, size, 0);
 
-#    if defined(__unix__)
-    // The maximum permitted value for the getentropy length argument.
-    static constexpr size_t getentropy_length_limit = 256;
-    auto iterations = bytes.size() / getentropy_length_limit;
+        if (ret == -1 && errno != EINTR) [[unlikely]]
+            return Error::from_errno(errno);
 
-    for (size_t i = 0; i < iterations; ++i) {
-        if (getentropy(bytes.data(), getentropy_length_limit) != 0) {
-            fill_with_random_fallback();
-            return;
+        // Per the manual:
+        //
+        // > On success, getrandom() returns the number of bytes that were
+        // > copied to the buffer buf.  This may be less than the number of
+        // > bytes requested via buflen if either GRND_RANDOM was specified in
+        // > flags and insufficient entropy was present in the random source or
+        // > the system call was interrupted by a signal.
+        //
+        // > The user of getrandom() *must* always check the return value, to
+        // > determine whether either an error occurred or fewer bytes than
+        // > requested were returned.
+        if (ret > 0) [[likely]] {
+            size -= ret;
+            buf += ret;
         }
-
-        bytes = bytes.slice(getentropy_length_limit);
     }
+#elif defined(AK_OS_WINDOWS)
+    if (size > NumericLimits<u32>::max()) [[unlikely]]
+        return Error::from_string_literal("The input is too large");
```

Comment:
This looks like a footgun: either we should support `size_t` on all plaforms (e.g. by handling this Windows case in chunks) or we should accept a `u32` so we don't end up with cross-platform issues if we're ever requesting bizarre amounts of random data.